common/plugin_common.h

```diff
@@ -109,9 +109,12 @@ PLUG_API int sasl_canonuser_init(const sasl_utils_t *utils, \
                                      plug, plugname); \
 }
 
-/* note: msg cannot include additional variables, so if you want to
+/* Only set the error message if one is not already set. This prevents overwriting
+ * plugin-supplied error messages with library error messages.
+ * note: msg cannot include additional variables, so if you want to
  * do a printf-format string, then you need to call seterror yourself */
-#define SETERROR( utils, msg ) (utils)->seterror( (utils)->conn, 0, (msg) )
+#define SETERROR( utils, msg ) \
+    if (!sasl_errdetail((utils)->conn)) (utils)->seterror( (utils)->conn, 0, (msg) )
```

Comment:
I am worried that sasl_errdetail() returns non NULL value whether or not the error was set on the connection. I need to test whether sasl_errdetail() is the right thing here, it might not be.

Reply:
I'm very new to SASL, so apologies if this is obvious question, but a freshly initiliazed server context returning a non-NULL value via sasl_errdetail() would not be inline with the documented return values: https://github.com/cyrusimap/cyrus-sasl/blob/master/docsrc/sasl/reference/manpages/library/sasl_errdetail.rst#return-value

That doesn't mean that it isn't happening, but it does raise a separate concern.

I'll work on adding some test cases to this patch and in doing so, I will address the concern that you've raised.